### PR TITLE
Globally upgrade to syn 2.x and latest quote and proc_macro2 1x versions

### PR DESCRIPTION
**Pull request by sam0x17**
polkadot companion: https://github.com/paritytech/polkadot/pull/7022

## Description

We have a lot of different versions of dtolnay's quote/syn/proc-macro2 crates floating around. Dtolnay recently came out with syn 2.x

This PR sets all versions of proc-macro2, syn, and quote to the versions listed below, which at the time of writing are the latest 1.x version for each crate (or 2.x in the case of syn)

This should improve compile-times since only one version of these crates will be compiled globally, and will also reduce confusing edge cases due to version differences

## Crates Upgraded

proc-macro2 => 1.0.56
syn => 2.0.14
quote => 1.0.26

### CODE REVIEW

**Comment by sam0x17** [comment]
converting to draft because it looks like we might need some companions

**Comment by sam0x17** [comment]
cumulus is already on 2.x and the latest for the other crates so no need for a cumulus companion

**Review by KiChjang — APPROVED** [review]

**Comment by bkchr** [comment]
> cumulus is already on syn 2.x and the latest for the other crates so no need for a cumulus companion

Then not updating syn to 2.x makes no sense.

**Comment by bkchr** [comment]
Could you please just update to 2.x for syn?

In general, cargo unifies across patch/minor releases which you already see with your pr here. You changed multiple Cargo.toml files, but the Cargo.lock wasn't changed. This means that we already used these versions.

**Comment by sam0x17** [comment]
> Could you please just update to 2.x for syn?
> 
> In general, cargo unifies across patch/minor releases which you already see with your pr here. You changed multiple Cargo.toml files, but the Cargo.lock wasn't changed. This means that we already used these versions.

Yes, let me do that

**Comment by sam0x17** [comment]
actually @bkchr it seems to be non-trivial. Upgrading to syn 2.x I get a ton of errors on substrate. For some reason cumulus is fine with it though

**Comment by koute** [comment]
Yes, we should update to `syn` 2, but it might not be just a trivial version bump as [there are breaking API changes between v1 and v2](https://github.com/dtolnay/syn/releases/tag/2.0.0).

@sam0x17 From a practical point of view it'd be the best to just update the deps to v2 one-by-one and fix the errors as you go.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658303

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658287

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658298

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658302

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658288

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658304

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658358

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658342

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658343

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658353

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658357

**Comment by bkchr** [comment]
> actually @bkchr it seems to be non-trivial. Upgrading to syn 2.x I get a ton of errors on substrate. For some reason cumulus is fine with it though

Cumulus probably proc-macro isn't that complex and maybe uses an api that didn't changed!

File: primitives/api/proc-macro/src/decl_runtime_apis.rs [hunk 1]
```diff
@@ -178,35 +179,27 @@ fn generate_versioned_api_traits(
...
-
-				Ok((old_name, version))
-			},
+	renamed.parse_nested_meta(|meta| {
+		let old_name_lit: LitStr = meta.input.parse()?;
+		old_name = Some(old_name_lit.value());
+		let _comma: Comma = meta.input.parse()?;
+		let version_lit: LitInt = meta.input.parse()?;
+		version = Some(version_lit.base10_parse()?);
+		Ok(())
```
Comment (bkchr) [inline primitives/api/proc-macro/src/decl_runtime_apis.rs:199]: We should return an error if there are tokens left as it was done before.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658414

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658413

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658398

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658399

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2658409

**Review by bkchr — APPROVED** [review]

**Review by koute — APPROVED** [review]

**Comment by sam0x17** [comment]
note, will need to get https://github.com/drahnr/expander/pull/6 merged if we want to do syn 2.x

**Comment by bkchr** [comment]
> note, will need to get [drahnr/expander#6](https://github.com/drahnr/expander/pull/6) merged if we want to do syn 2.x

`scale-codec` also doesn't yet use `syn` 2.0. There are also probably plenty of other crates that are not yet ported. I don't see a reason to wait.

**Comment by sam0x17** [comment]
@bkchr I can't get polkadot to build without upgrading expander's syn. There is a very subtle conflict on `syn::Error` not be compatible between the latest 1.x and the latest 2.x, if you look at my latest push

**Comment by bkchr** [comment]
There is no commit on the companion.

**Comment by sam0x17** [comment]
This one causes the failure https://github.com/paritytech/polkadot/pull/7022/commits/1c63df999ab0049acaa4651d0038e75705e0b90f

**Comment by sam0x17** [comment]
This is the error @bkchr, took me a good bit to figure out that it is because of syn versions:

image

**Comment by sam0x17** [comment]
So he accepted my PR, though no new release has been cut yet. I just pushed up a test commit manually updating the version of expander to use the latest git just to confirm that CI passes with it. Will remove that once he cuts a new release

**Comment by sam0x17** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2672578

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2672577

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2672576

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2672572

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2672566

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2672562

**Comment by KiChjang** [comment]
> So he accepted my PR, though no new release has been cut yet. I just pushed up a test commit manually updating the version of expander to use the latest git just to confirm that CI passes with it. Will remove that once he cuts a new release

You don't really need to wait for him to publish -- you said before that `paritytech-devs` are owners of his crate on crates.io as well, correct? If so, you can just run `cargo publish` locally with the new version, and it will Just Work.

**Comment by sam0x17** [comment]
> > So he accepted my PR, though no new release has been cut yet. I just pushed up a test commit manually updating the version of expander to use the latest git just to confirm that CI passes with it. Will remove that once he cuts a new release
> 
> You don't really need to wait for him to publish -- you said before that `paritytech-devs` are owners of his crate on crates.io as well, correct? If so, you can just run `cargo publish` locally with the new version, and it will Just Work.

he published an hour after he accepted my PR apparently so we good

**Comment by sam0x17** [comment]
bot merge

**Pull request merged**